### PR TITLE
applications: nrf5340_audio: Implement sync led

### PR DESCRIPTION
**Pull request by Khuongh**
Implementing synchronization for led1_blue on headsets

Signed-off-by: Khuong Huynh <khuong.huynh@nordicsemi.no>

### CODE REVIEW

**Comment by Khuongh** [comment]
New rebase:

Fixed wrapping for when seq_num goes over 65535, such that the it doesn't miss that blink. And added offset for when wrapping happens. 

The Headsets are not receiving the same sequence number in CIS. Have tried to send the same value from the GW, but it seems that the sequence number are received/incremented from somewhere else. Talked to Erik about this and he is going to look further into it.

File: applications/nrf5340_audio/src/audio/audio_datapath.c [hunk 1]
```diff
@@ -747,6 +756,34 @@ static void audio_datapath_i2s_stop(void)
...
+			sync_led_last_seq_num += (SYNC_LED_SEQ_NR_PERIOD / 2);
+			return true;
+		}
+	}
+	return false;
+}
+
+static bool sync_led_toggle_check(uint32_t seq_num, uint32_t offset)
+{
+	if ((seq_num % SYNC_LED_SEQ_NR_PERIOD) - offset == SYNC_LED_SEQ_NR_ON) {
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/audio_datapath.c:780]: It seems that we can run into a missed seq_num issue here as well. If this if is meant to return true at e.g. `seq_num` = 3000, but if we miss 3000 and get 3001 instead, it should still turn on?

File: applications/nrf5340_audio/src/audio/audio_sync_timer.c [hunk 2]
```diff
@@ -79,6 +86,89 @@ static void event_handler(nrf_timer_event_t event_type, void *ctx)
...
+	if (nrfx_gpiote_is_init()) {
+		LOG_DBG("GPIOTE is already initiated");
+	} else {
+		ret = nrfx_gpiote_init(NRFX_GPIOTE_DEFAULT_CONFIG_IRQ_PRIORITY);
+		if (ret - NRFX_ERROR_BASE_NUM) {
+			LOG_ERR("Failed to init GPIOTE (err: %d).", ret);
+			return -EIO;
+		}
+
+		nrfx_gpiote_global_callback_set((void *)nrfx_gpiote_handler, NULL);
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/audio_sync_timer.c:115]: Do we need to set a callback if it is not used?

File: modules/tfm/zephyr/CMakeLists.txt [hunk 3]
```diff
@@ -146,6 +146,11 @@ set_property(TARGET zephyr_property_target
     -DCRYPTO_ENGINE_BUF_SIZE=${CONFIG_TFM_CRYPTO_ENGINE_BUF_SIZE}
     -DCRYPTO_CONC_OPER_NUM=${CONFIG_TFM_CRYPTO_CONC_OPER_NUM}
     -DCRYPTO_IOVEC_BUFFER_SIZE=${CONFIG_TFM_CRYPTO_IOVEC_BUFFER_SIZE}
+    # Pass Zephyr Python to TF-M so both uses identical Python.
+    -DPython3_EXECUTABLE=${Python3_EXECUTABLE}
```
Comment (koffes) [inline modules/tfm/zephyr/CMakeLists.txt:150]: Why are these changes required?

**Comment by erikrobstad** [comment]
I am going to review the files, but I have one general comment:
The documentation should be updated, at least in:

- https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/applications/nrf5340_audio/README.html#leds
- https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/applications/nrf5340_audio/README.html#testing

File: applications/nrf5340_audio/src/audio/audio_sync_timer.c [hunk 4]
```diff
@@ -79,6 +86,89 @@ static void event_handler(nrf_timer_event_t event_type, void *ctx)
...
+		LOG_WRN("Invalid audio sync timer instance");
+		return -ECANCELED;
+	}
+
+	if (nrfx_gpiote_is_init()) {
+		LOG_DBG("GPIOTE is already initiated");
+	} else {
+		ret = nrfx_gpiote_init(NRFX_GPIOTE_DEFAULT_CONFIG_IRQ_PRIORITY);
+		if (ret - NRFX_ERROR_BASE_NUM) {
+			LOG_ERR("Failed to init GPIOTE (err: %d).", ret);
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/audio/audio_sync_timer.c:111]: Nitpick:

```suggestion
			LOG_ERR("nrfx GPIOTE init error - Return value: %d", ret);
```

We usually dont have periods in logs

File: applications/nrf5340_audio/src/audio/audio_sync_timer.c [hunk 5]
```diff
@@ -79,6 +86,89 @@ static void event_handler(nrf_timer_event_t event_type, void *ctx)
...
+		nrfx_gpiote_global_callback_set((void *)nrfx_gpiote_handler, NULL);
+		LOG_DBG("GPIOTE initialized");
+	}
+
+	/* GPIOTE out configs for sync LED */
+	nrfx_gpiote_out_config_t const gpiote_cfg = NRFX_GPIOTE_CONFIG_OUT_TASK_HIGH;
+
+	ret = nrfx_gpiote_out_init(SYNC_LED, &gpiote_cfg);
+	if (ret - NRFX_ERROR_BASE_NUM) {
+		LOG_ERR("nrfx GPIOTE init error - Return value: %d", ret);
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/audio/audio_sync_timer.c:124]:
```suggestion
		LOG_ERR("nrfx GPIOTE out init error - Return value: %d", ret);
```

File: applications/nrf5340_audio/src/audio/audio_sync_timer.c [hunk 6]
```diff
@@ -112,7 +209,9 @@ static int audio_sync_timer_init(const struct device *unused)
 		LOG_ERR("nrfx DPPI channel alloc error (I2S frame start) - Return value: %d", ret);
 		return ret;
 	}
-	nrf_timer_subscribe_set(timer_instance.p_reg, AUDIO_SYNC_TIMER_I2S_FRAME_START_EVT_CAPTURE,
+	nrf_timer_event_t audio_sync_timer_i2s_frame_start_capture_task =
```
[the remark this answers is not in the document]
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/audio_sync_timer.c:212]: Also on line 228 (not a part of your changes, but would be nice to just fix it)

**Review by erikrobstad — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/audio/audio_datapath.c (hunk 1 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/audio/audio_datapath.c]: Added new suggestion

File: applications/nrf5340_audio/src/audio/audio_sync_timer.c (hunk 2 above)
Comment (Khuongh) [reply applications/nrf5340_audio/src/audio/audio_sync_timer.c]: Tested the functionality without this callback, with running uinit_gipote and everything worked. Seems like this callback is not needed. Removed from code.

File: applications/nrf5340_audio/src/audio/audio_datapath.c [hunk 7]
```diff
@@ -123,6 +123,11 @@ LOG_MODULE_REGISTER(audio_datapath, CONFIG_LOG_AUDIO_DATAPATH_LEVEL);
 /* How often to print underrun warning */
 #define UNDERRUN_LOG_INTERVAL_BLKS 5000
 
+#define SYNC_LED_PERIOD_US 640000
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/audio_datapath.c:126]: Many great changes. Though, we cannot have this define, as we will run into offset issues if we do this. 
E.g. if CONFIG_AUDIO_FRAME_DURATION_US = 7500, SYNC_LED_SEQ_NR_PERIOD = 640000: SYNC_LED_SEQ_NR_PERIOD would be 85.33

I would remove this line, and instead just have:
`#define SYNC_LED_SEQ_NR_PERIOD 64` /* Value must be 2^N*/`

File: applications/nrf5340_audio/src/audio/audio_datapath.h [hunk 8]
```diff
@@ -123,11 +123,12 @@ void audio_datapath_sdu_ref_update(uint32_t sdu_ref_us);
  * @param buf Pointer to audio data frame
  * @param size Size of audio data frame in bytes
  * @param sdu_ref_us ISO timestamp reference from BLE controller
+ * @param seq_num ISO packet sequence number of the first fragment in the SDU
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/audio_datapath.h:126]: CI gives an error on the license, but that will be fixed if you rebase.

File: applications/nrf5340_audio/README.rst [hunk 9]
```diff
@@ -685,6 +685,7 @@ To indicate the tasks performed, the application uses the LED behavior described
 |                          | Blinking blue - Depending on the device and the mode:                                               |
 |                          |                                                                                                     |
 |                          | * Headset: Kits have started streaming audio (BIS and CIS modes).                                   |
+|                          |            LEDS will blink in sync to visualize synchronization of audio between kits.              |
```
Comment (koffes) [inline applications/nrf5340_audio/README.rst:688]: Nitpick: LEDs. We need to discuss on Monday - ref. the seq_num issue. We may have to document differently until that has been resolved.

**Comment by koffes** [comment]
As discussed internally: We would love to get this feature in, but the LE Audio standard does not support the current method as-is. We are following up with the standard, and hopefully, we can open this again later.
